### ISSUE TITLE
Add AV1 Support

### ISSUE THREAD

**Issue by billylindeman** [webrtc-rs/webrtc#193]
Pion added av1 support:

https://github.com/pion/webrtc/pull/2071
https://github.com/pion/webrtc/pull/2176

**Comment by k0nserv** [comment]
This would be nice to have, but I don't expect it to be priority for those working on the project at the moment. 

If it's something you are interested, it'd be great if you could contribute it

**Comment by billylindeman** [comment]
@k0nserv Absolutely!  I did some experimenting with just supporting the mimetype / codec to get AV1 flowing and it was pretty simple. Going to post a PR for that shortly.  

Looking to eventually add support for the dependency descriptor rtp header extension (AV1 SVC) but don't have the cycles for it at the moment =] 

Reference: https://aomediacodec.github.io/av1-rtp-spec/

**Comment by shiqifeng2000** [comment]
Similar to H265 support

**Comment by harrynull** [comment]
I wonder what the current state of AV1 support is. I added the codec but apparently I also need a payloader for it?

**Comment by k0nserv** [comment]
It depends. If you're only forwarding RTP packets there should be no need for a payloader, but if you're using `TrackLocalStaticSample`, then yes a payloader is needed. 

If you want to contribute it that would be very welcome, the details on how to are available in [RTP Payload Format For AV1 (v1.0)](https://aomediacodec.github.io/av1-rtp-spec/)

**Comment by juancampa** [comment]
I'm taking a stab at implementing the AV1 payloader today. My plan is to base it on [pion's payloader](https://github.com/pion/rtp/blob/master/codecs/av1_packet.go#LL31C1-L31C79). Any help or guidelines would be greatly appreciated.

**Comment by juancampa** [comment]
I'm trying to understand the scope of this work. Do you think that implementing the AV1 payloader and returning it from `payloader_for_codec` would be enough to stream AV1 using `TrackLocalStaticSample`? Or, is there anything else that I would need to build?

**Comment by juancampa** [comment]
An update on this: I ported pion's payloader but Chrome is still not happy with it, it keeps sending PILs and not displaying anything. I have proof that the raw encoded frames are correct (I dumped it into a file and I can `ffplay` it). Still trying to figure out why Chrome is not able to decode the stream though.

I'm going to step debug a local Chromium build to investigate but any tips on how to debug the decoder side would be helpful.

**Comment by juancampa** [comment]
@k0nserv do you know if I have to implement something for the "packetization" too?

**Comment by k0nserv** [comment]
I'm not really active in this project anymore, that said no you should only need packetization. It looks like you've made some mistake with a length field, e.g. a an overflow/underflow, wrong endianess, or something with the offsets in the RTP payload being wrong

**Comment by HITGIF** [comment]
Hi all, I've fleshed out an implementation of the AV1 payloader by porting over [Chromium's C++ implementation](https://chromium.googlesource.com/external/webrtc/+/4e513346ec56c829b3a6010664998469fc237b35/modules/rtp_rtcp/source/rtp_packetizer_av1.cc) in this PR: https://github.com/webrtc-rs/webrtc/pull/471.

I believe the problem with @juancampa 's approach is that [pion's payloader](https://github.com/pion/rtp/blob/master/codecs/av1_packet.go?rgh-link-date=2023-05-16T15%3A38%3A05Z#LL31C1-L31C79) it assumes the `payload` always contains exactly one OBU, which is apparently not true. The Chromium's implementation handles parsing multiple OBUs from the `payload` and then packetizing accordingly.

The current implementation does have some caveats as noted in the PR description, such as potentially sending (marking) redundant key frames. However, I have tested (as being used in [my own project](https://github.com/mira-screen-share/sharer)) that the packets can be successfully decoded by Chrome with no apparent issues. Therefore I think this is a good start.

@k0nserv @juancampa @billylindeman Feel free to give it a try, and any feedback is welcomed :)

**Comment by sdroege** [comment]
> assumes the `payload` always contains exactly one OBU, which is apparently not true

This is mandated by the spec but Pion at least produces payloads with multiple OBUs at once.
